### ISSUE TITLE
RFE: add default() method to Form::Tiny::Form to ease creating a default value for a nested form

### ISSUE THREAD

**Issue by djerius** [bbrtj/perl-form-tiny#11]
If a field is a nested form with defaults, I haven't been able to come up with a straightforward approach to create a default for the field from the defaults for the nested form.

For example:

```
use v5.36;
use Data::Dumper;

package My::SubForm {
    use Form::Tiny;
    form_field file => ( default => sub { 'boo' } );
}

package My::Form {
    use Form::Tiny;
    use Data::Dumper;
    use List::Util 'first';

    form_field input =>
      ( type => My::SubForm->new,
        default => sub  ($self) {
            my $field  = (first { $_->name eq 'input' } $self->field_defs->@* ) // die "can't find input?";
            my $form = $field->type;
            # I'm sure there's a more efficient way to get the defaults
            $form->set_input({});
            $form->valid  or die Dumper $form->errors_hash;
            return $form->fields;
        },
    );
}

my $form = My::Form->new;
$form->set_input( {} );
$form->valid or die Dumper $form->errors_hash;
say Dumper $form->fields;
```

I can think of a couple things that would make it easier:

1. Pass the field definition to the default() subroutine (e.g. default( $self, $field )
2. Add a form method which returns the default form data

This would turn the default code above into

```
sub ( $self, $field ) { return $field->type->default }
```

**Comment by bbrtj** [comment]
Hello!

The problem is not really with the default fields. Nested form would fill them out just fine... if it was validated at all. As you may have noticed in the docs, there's a special case for `adjust` when dealing with subforms: https://metacpan.org/dist/Form-Tiny/view/lib/Form/Tiny/Manual.pod#Nested-forms

However, the field is never adjusted (or validated at all) if there is no input for it. This is why you see no defaults. I agree this may be a bit confusing.

The easiest fix to your problem is by adding a hook in the main form:
```perl
form_hook before_validate => sub ($self, $input) {
    $input->{input} //= {};
};
```

This way there is always something to validate, which then causes adjustment, injecting default fields from the subform:
```
$VAR1 = {
          'input' => {
                       'file' => 'boo'
                     }
        };
```

This is pretty straightforward. I haven't yet made up my mind whether this should be the default behavior, just like the special adjustment case I linked above. Surely something that should at least end up in Manual::Cookbook.

**Comment by bbrtj** [comment]
I think the least surprising behavior would be to just specify the default this way:
```perl
sub ($self) { return {}; }
```

This would work if Form::Tiny would be changed to also validate the default (which it does not currently). I think it is a reasonable thing to do, with three small problems associated with it:
- it breaks backcompat (although can be considered a fix - if you set a default which would not pass the validation, you have a bug)
- it would affect performance slightly
- it would report invalid default value late (during validation), because validation can't be run before instantiating the form object

Note that if the subform has any required fields, the default would have to take them into account:
```perl
sub ($self) { return {required => 'anything'}; }
```

(interestingly, required fields in subform would not currently cause validation error if the subform was not itself required in the main form - something worth noting in the cookbook as well)

Edit: However, automatically adjusting the default seems like a no-no. This requires more thought.

**Comment by djerius** [comment]
Hi,

> As you may have noticed in the docs, there's a special case for adjust when dealing with subforms: https://metacpan.org/dist/Form-Tiny/view/lib/Form/Tiny/Manual.pod#Nested-forms
> 
> However, the field is never adjusted (or validated at all) if there is no input for it. This is why you see no defaults. 

Ahh. I hadn't realize that it was `adjust` which accomplished this behavior; as long as a value was passed for the subform everything worked as I expcted to, so I didn't look deeper into the mechanism.

> The easiest fix to your problem is by adding a hook in the main form:
> 
> ```
> form_hook before_validate => sub ($self, $input) {
>     $input->{input} //= {};
> };
> ```
> 
> This way there is always something to validate, which then causes adjustment, injecting default fields from the subform:
>

Much cleaner than my approach!  However, I prefer your later idea

> I think the least surprising behavior would be to just specify the default this way:
> 
> `sub ($self) { return {}; }`
> 

because it keeps the full specification of the field within the `form_field` stanza, rather than splitting it into two places.

**Comment by bbrtj** [comment]
I have implemented validating of the default value if the field is a subform - this seems like a good change. No adjustment or anything, simple validation of default and taking its fields. I used similar logic for adjustments - no more special case hacks! All is well documented, you can check the new documentation in the commit.

It may need some ironing, but should make what you're trying to do much easier. Thanks for reporting. I will release to CPAN in a couple of days tops.

**Comment by bbrtj** [comment]
@djerius this was released in 2.18. It's documented in [the manual](https://metacpan.org/release/BRTASTIC/Form-Tiny-2.18_01/view/lib/Form/Tiny/Manual.pod#Nested-forms)

> default value for the subform must be its valid input
>
> It will be used to determine the default values for fields of the subform. As with regular default values, it will not use filters, coercions or adjustments from the main form field.
